### PR TITLE
[`T5`] Fix torchquant issue

### LINKED ISSUE #21839: quantize_dynamic on T5 model results in `AttributeError: 'function' object has no attribute 'dtype'`
### System Info

- `transformers` version: 4.26.1
- Platform: Linux-5.10.147+-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.12.1
- PyTorch version (GPU?): 1.13.1+cu116 (False)
- Tensorflow version (GPU?): 2.11.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@younesbelkada

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Paste the following into colab:

```
!pip install transformers sentencepiece accelerate
import torch
from transformers import T5ForConditionalGeneration, T5Tokenizer
model_name = "google/flan-t5-small"
tokenizer = T5Tokenizer.from_pretrained(model_name)
model = T5ForConditionalGeneration.from_pretrained(model_name)
model = torch.quantization.quantize_dynamic(model, {torch.nn.Linear}, dtype=torch.qint8)
input_text = "Answer the following yes/no question by reasoning step-by-step. Can you write a whole Haiku in a single tweet?"
input_ids = tokenizer(input_text, return_tensors="pt").input_ids
output = model.generate(input_ids)
``` 

### Expected behavior

No Error

# Bug
```
AttributeError                            Traceback (most recent call last)
[<ipython-input-3-96b349bbc122>](https://localhost:8080/#) in <module>
      1 input_text = "Answer the following yes/no question by reasoning step-by-step. Can you write a whole Haiku in a single tweet?"
      2 input_ids = tokenizer(input_text, return_tensors="pt").input_ids
----> 3 output = model.generate(input_ids)

10 frames
[/usr/local/lib/python3.8/dist-packages/transformers/models/t5/modeling_t5.py](https://localhost:8080/#) in forward(self, hidden_states)
    314         # See https://github.com/huggingface/transformers/issues/20287
    315         # we also make sure the weights are not in `int8` in case users will force `_keep_in_fp32_modules` to be `None``
--> 316         if hidden_states.dtype != self.wo.weight.dtype and self.wo.weight.dtype != torch.int8:
    317             hidden_states = hidden_states.to(self.wo.weight.dtype)
    318 

AttributeError: 'function' object has no attribute 'dtype'
```

# Fix

When I commented out the lines 316-317 in `transformers/models/t5/modeling_t5.py`, the model runs.

`quantize_dynamic` converts `self.wo.weight` into a bound function which when called, returns the weights. Hence `self.wo.weight` is a function with no attribute `dtype`.

Bug is introduced by #21281

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes #21839

This PR fixes a bug that was introduced with https://github.com/huggingface/transformers/pull/21281 - before this PR, the snippet below was working: 
```python
import torch
from transformers import T5ForConditionalGeneration, T5Tokenizer
model_name = "google/flan-t5-small"
tokenizer = T5Tokenizer.from_pretrained(model_name)
model = T5ForConditionalGeneration.from_pretrained(model_name)
model = torch.quantization.quantize_dynamic(model, {torch.nn.Linear}, dtype=torch.qint8)
input_text = "Answer the following yes/no question by reasoning step-by-step. Can you write a whole Haiku in a single tweet?"
input_ids = tokenizer(input_text, return_tensors="pt").input_ids
output = model.generate(input_ids)
```
On the `main` branch the snippet does not work anymore due to the check `self.wo.weight.dtype` since ` torch.quantization.quantize_dynamic` converts all `nn.Linear` layers to a bound function, leading to an error.
Since the users were able to run this snippet on previous versions, I think that we should support this feature.

Added also a cool test for that

cc @sgugger

### CODE REVIEW

File: tests/models/t5/test_modeling_t5.py [hunk 1]
```diff
@@ -880,6 +880,19 @@ def model(self):
...
+        Test that a simple `torch.quantization.quantize_dynamic` call works on a T5 model.
+        """
+        model_name = "google/flan-t5-small"
+        tokenizer = T5Tokenizer.from_pretrained(model_name)
+        model = T5ForConditionalGeneration.from_pretrained(model_name)
+        model = torch.quantization.quantize_dynamic(model, {torch.nn.Linear}, dtype=torch.qint8)
+        input_text = "Answer the following yes/no question by reasoning step-by-step. Can you write a whole Haiku in a single tweet?"
+        input_ids = tokenizer(input_text, return_tensors="pt").input_ids
+        _ = model.generate(input_ids)
+
```
Comment (ArthurZucker) [inline tests/models/t5/test_modeling_t5.py:895]: Nice test 😉

**Review by ArthurZucker — APPROVED** [review]
LGTM thanks for fixing this

**Review by sgugger — APPROVED** [review]
Thanks for the fix!

**Comment by younesbelkada** [comment]
Can confirm t5 & int8 slow tests are passing, merging!

**Pull request merged**